src/simulation.cpp

```diff
@@ -464,74 +463,61 @@ void initialize_history(Particle& p, int64_t index_source)
     auto site = sample_external_source(&seed);
     p.from_source(&site);
   }
-  p.current_work_ = index_source;
+  p.current_work() = index_source;
 
   // set identifier for particle
-  p.id_ = simulation::work_index[mpi::rank] + index_source;
+  p.id() = simulation::work_index[mpi::rank] + index_source;
 
   // set progeny count to zero
-  p.n_progeny_ = 0;
+  p.n_progeny() = 0;
 
   // Reset particle event counter
-  p.n_event_ = 0;
+  p.n_event() = 0;
 
   // set random number seed
-  int64_t particle_seed = (simulation::total_gen + overall_generation() - 1)
-    * settings::n_particles + p.id_;
-  init_particle_seeds(particle_seed, p.seeds_);
+  int64_t particle_seed =
+    (simulation::total_gen + overall_generation() - 1) * settings::n_particles +
+    p.id();
+  init_particle_seeds(particle_seed, p.seeds());
 
   // set particle trace
-  p.trace_ = false;
+  p.trace() = false;
   if (simulation::current_batch == settings::trace_batch &&
       simulation::current_gen == settings::trace_gen &&
-      p.id_ == settings::trace_particle) p.trace_ = true;
+      p.id() == settings::trace_particle)
+    p.trace() = true;
 
   // Set particle track.
-  p.write_track_ = false;
+  p.write_track() = false;
   if (settings::write_all_tracks) {
-    p.write_track_ = true;
+    p.write_track() = true;
   } else if (settings::track_identifiers.size() > 0) {
     for (const auto& t : settings::track_identifiers) {
       if (simulation::current_batch == t[0] &&
-          simulation::current_gen == t[1] &&
-          p.id_ == t[2]) {
-        p.write_track_ = true;
+          simulation::current_gen == t[1] && p.id() == t[2]) {
+        p.write_track() = true;
         break;
       }
     }
   }
 
   // Display message if high verbosity or trace is on
-  if (settings::verbosity >= 9 || p.trace_) {
-    write_message("Simulating Particle {}", p.id_);
+  if (settings::verbosity >= 9 || p.trace()) {
+    write_message("Simulating Particle {}", p.id());
   }
 
   // Add paricle's starting weight to count for normalizing tallies later
   #pragma omp atomic
-  simulation::total_weight += p.wgt_;
-
-  initialize_history_partial(p);
-}
+  simulation::total_weight += p.wgt();
 
-void initialize_history_partial(Particle& p)
-{
   // Force calculation of cross-sections by setting last energy to zero
   if (settings::run_CE) {
-    for (auto& micro : p.neutron_xs_) micro.last_E = 0.0;
+    p.invalidate_neutron_xs();
   }
 
   // Prepare to write out particle track.
-  if (p.write_track_) add_particle_track(p);
-
-  // Every particle starts with no accumulated flux derivative.
```

Comment:
Yeah, again, they're in the constructor for `ParticleData` now since this code differs between SOA and non-SOA.

Reply:
I can see that this change works fine for history-based mode, where each particle is constructed and destroyed each power iteration. However, in event-based mode, the particles persist between iterations such that the constructors are only called once per simulation, not once per iteration. Does this cause issues?

Reply:
Yeah, so, look at the constructor of ParticleData. It's in there. It specifically checks if we're in event mode, then sets things up so that it'll be good for the rest of the simulation. As for the track writing thing directly near your comment here, it's still in initialize_history, just not in inititalize_history_partial, which seemed to only be used in one place and had data-layout-dependent code so I just lumped what isn't now in the ctor of ParticleData into initialize_history.

Reply:
The way it currently is in OpenMC, the particle's `flux_derivs_` are set to zero when `initialize_history()` is called (i.e., every batch). By moving that to the particle constructor, the `flux_derivs_` now only get set to zero once at program initialization in event-based mode, not once per iteration. Does this cause issues?

Reply:
Nope. There is now a method called `zero_flux_derivs`. It gets called in Particle's `from_source`, so the behavior should be the same.